shell/browser/ui/devtools_manager_delegate.cc

```diff
@@ -102,6 +106,20 @@ void DevToolsManagerDelegate::HandleCommand(
     content::DevToolsAgentHostClientChannel* channel,
     base::span<const uint8_t> message,
     NotHandledCallback callback) {
+  crdtp::Dispatchable dispatchable(crdtp::SpanFrom(message));
+  DCHECK(dispatchable.ok());
+  if (crdtp::SpanEquals(crdtp::SpanFrom(kBrowserCloseMethod),
+                        dispatchable.Method())) {
+    // In theory, we should respond over the protocol saying that the
+    // Browser.close was handled. But doing so requires instantiating the
+    // protocol UberDispatcher and generating proper protocol handlers.
+    // Since we only have one method and it is supposed to close Electron,
+    // we don't need to add this complexity. Should we decide to support
+    // metohds like Browser.setWindowBounds, we'll need to do it though.
+    base::PostTask(FROM_HERE, {content::BrowserThread::UI},
+                   base::BindOnce([]() { Browser::Get()->Shutdown(); }));
```

Comment:
Based on the protocol documentation, this is supposed to be a graceful shutdown but `Browser::Get()->Shutdown()` isn't graceful when there are multiple windows with embedded `webContents`. Ideally this should be `Browser::Get()->Quit()` but it has handlers on the app side that are allowed to prevent shutdown. Also by graceful, is the window allowed to run its unload handlers ?

Reply:
In `Browser.close` graceful means not generating the core dump. I.e. browser initiates the exit rather than process is killed. No unload handlers are allowed.

Reply:
Thanks for the context! But in chrome's case I see they are trying to force close any open browser windows https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/lifetime/application_lifetime.cc;l=302-323 . Because quitting the message loop before cleaning up the windows seems like a path for crash, especially when there are user scripts running in the main process, simple demo https://gist.github.com/5ecb1eb5a7dca08d2681ab18595f9a45 , in the local electron build change `app.quit` to call `Browser::Get()->Shutdown()` , the app will core dump.

Reply:
Ah, got it! Ok, in this case let me call `Browser::Get()->Quit()` to keep things simple. If we need to force no handlers on the app side, we can follow up. Thanks for the hints!

Reply:
Sounds good, thanks!